chain/get_ancestors.go

```diff
@@ -66,76 +71,50 @@ func GetRecentAncestors(ctx context.Context, base types.TipSet, chainReader Read
-	// be reused.
-	tsas, err := chainReader.GetTipSetAndState(ctx, firstExtraRandomnessAncestorsCids.String())
+	tsas, err := chainReader.GetTipSetAndState(firstExtraRandomnessAncestorsCids)
 	if err != nil {
 		return nil, err
 	}
-	historyCh = chainReader.BlockHistory(ctx, tsas.TipSet)
-	extraRandomnessAncestors, err := CollectAtMostNTipSets(ctx, historyCh, lookback)
+	iterator = IterAncestors(ctx, chainReader, tsas.TipSet)
+	extraRandomnessAncestors, err := CollectAtMostNTipSets(ctx, iterator, lookback)
 	if err != nil {
 		return nil, err
 	}
 	return append(provingPeriodAncestors, extraRandomnessAncestors...), nil
 }
 
 // CollectTipSetsOfHeightAtLeast collects all tipsets with a height greater
-// than or equal to minHeight from the input channel.  Precondition, the input
-// channel contains interfaces which may be tipsets or errors.
-func CollectTipSetsOfHeightAtLeast(ctx context.Context, ch <-chan interface{}, minHeight *types.BlockHeight) ([]types.TipSet, error) {
+// than or equal to minHeight from the input tipset.
+func CollectTipSetsOfHeightAtLeast(ctx context.Context, iterator *TipsetIterator, minHeight *types.BlockHeight) ([]types.TipSet, error) {
 	var ret []types.TipSet
-	for {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		case raw, more := <-ch:
-			if !more {
-				return ret, nil
-			}
-			switch raw := raw.(type) {
-			case error:
-				return nil, raw
-			case types.TipSet:
-				// Add tipset to ancestors.
-				h, err := raw.Height()
-				if err != nil {
-					return nil, err
-				}
-
-				// Check for termination.
-				if types.NewBlockHeight(h).LessThan(minHeight) {
-					return ret, nil
-				}
-				ret = append(ret, raw)
-			}
+	var err error
+	var h uint64
+	for ; !iterator.Complete(); err = iterator.Next() {
+		if err != nil {
+			return nil, err
 		}
+		h, err = iterator.Value().Height()
+		if err != nil {
+			return nil, err
+		}
+		if types.NewBlockHeight(h).LessThan(minHeight) {
+			return ret, nil
+		}
+		ret = append(ret, iterator.Value())
 	}
+	return ret, nil
 }
 
 // CollectAtMostNTipSets collect N tipsets from the input channel.  If there
 // are fewer than n tipsets in the channel it returns all of them.
-func CollectAtMostNTipSets(ctx context.Context, ch <-chan interface{}, n uint) ([]types.TipSet, error) {
+func CollectAtMostNTipSets(ctx context.Context, iterator *TipsetIterator, n uint) ([]types.TipSet, error) {
 	var ret []types.TipSet
-	for i := uint(0); i < n; i++ {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		case raw, more := <-ch:
-			if !more {
-				return ret, nil
-			}
-			switch raw := raw.(type) {
-			case error:
-				return nil, raw
-			case types.TipSet:
-				// Add tipset to ancestors.
-				ret = append(ret, raw)
-			}
+	var err error
+	for i := uint(0); i < n && !iterator.Complete(); i++ {
+		ret = append(ret, iterator.Value())
+		if err = iterator.Next(); err != nil {
+			return nil, err
```

Comment:
Thanks, this diff is a compelling argument for iterators as a pattern in preference to channels.